metagraph/src/graph/representation/succinct/dbg_succinct_cached.hpp

```diff
@@ -0,0 +1,362 @@
+#ifndef __DBG_SUCCINCT_CACHED__
+#define __DBG_SUCCINCT_CACHED__
+
+#include <cache.hpp>
+#include <lru_cache_policy.hpp>
+
+#include "boss.hpp"
+#include "dbg_succinct.hpp"
+#include "kmer/kmer_extractor.hpp"
+
+namespace mtg {
+namespace graph {
+
+
+class DBGSuccinctCached : public DeBruijnGraph {
+  public:
+    typedef boss::BOSS::edge_index edge_index;
+    typedef boss::BOSS::TAlphabet TAlphabet;
```

Comment:
Why don't we make it a template class?

Or, does this class use something specific to DBGSuccinct, i.e., sentinel characters?

Reply:
I had a description for this in the pull request, but I'll also add a comment here.

This class caches the results for two things: 1) `get_node_sequence()`, and 2) `get_minus_k_value(k - 2)`

So, it has to be specific to DBGSuccinct. One could split this up such that a general class handles 1) and a more specific class handles 2), but I don't think it's worth implementing at this time.

Reply:
Why do you need to cache `get_minus_k_value(k - 2)`? Are you sure you don't cache the same nodes in both, and hence reconstruct the same nodes twice?

Reply:
`get_minus_k_value(k - 2)` returns two things: the decoded first character, and the corresponding BOSS node from the backwards traversal. So, the cache maps the input DBGSuccinct node to that BOSS node. It also caches the `bwd` of both of these, so the next time the function is called, it can fetch it from the cache.

The whole point of this is to speed up backwards traversal, so each step (on average) only needs to pay the cost of two `bwd` calls.

Reply:
Wouldn't caching `get_node_sequence()` be enough?

Reply:
Almost. If we modified `BOSS::get_node_seq` (or added a new helper function) to also return the last node after traversal, then we could have a single cache which keeps a pair of an encoded k-mer and that BOSS node.

Reply:
why just storing `get_node_sequence()` would not be enough with the current interface?

Reply:
Because that's only good enough for retrieving the value sequence of a node you've seen before. For any of its descendants, you'd need to recompute the sequence from scratch. If we have access to the last node that was accessed during the decoding, we could compute the full sequence of its descendent with a single `bwd` call (plus another `bwd` call to compute the corresponding key for the cache).

Reply:
I updated the implementation so it uses a single cache to store a pair with both the constructed node label, and the node resulting from the backwards traversals.